File: analytics/metrics-platform/src/test/java/org/wikimedia/metrics_platform/SamplingControllerTest.java
package org.wikimedia.metrics_platform;

import static org.assertj.core.api.Assertions.assertThat;
import static org.wikimedia.metrics_platform.config.sampling.SampleConfig.Identifier.DEVICE;
import static org.wikimedia.metrics_platform.config.sampling.SampleConfig.Identifier.SESSION;

import org.junit.jupiter.api.Test;
import org.wikimedia.metrics_platform.config.sampling.SampleConfig;
import org.wikimedia.metrics_platform.config.StreamConfig;
import org.wikimedia.metrics_platform.context.DataFixtures;

class SamplingControllerTest {

    private final SamplingController samplingController = new SamplingController(
            DataFixtures.getTestClientData(),
            new SessionController()
    );

    @Test void testGetSamplingValue() {
        double deviceVal = samplingController.getSamplingValue(DEVICE);
        assertThat(deviceVal).isBetween(0.0, 1.0);
    }

    @Test void testGetSamplingId() {
        assertThat(samplingController.getSamplingId(DEVICE)).isNotNull();
        assertThat(samplingController.getSamplingId(SESSION)).isNotNull();
    }

    @Test void testNoSamplingConfig() {
        StreamConfig noSamplingConfig = new StreamConfig("foo", "bar", null, null, null);
        assertThat(samplingController.isInSample(noSamplingConfig)).isTrue();
    }

    @Test void testAlwaysInSample() {
        StreamConfig alwaysInSample = new StreamConfig("foo", "bar", null,
                new StreamConfig.ProducerConfig(new StreamConfig.MetricsPlatformClientConfig(
                        null,
                        null,
                        null
                )),
                null
        );
        assertThat(samplingController.isInSample(alwaysInSample)).isTrue();
    }

    @Test void testNeverInSample() {
        StreamConfig neverInSample = new StreamConfig("foo", "bar", null,
                new StreamConfig.ProducerConfig(new StreamConfig.MetricsPlatformClientConfig(
                        null,
                        null,
                        null
                )),
                new SampleConfig(0.0, SESSION)
        );
        assertThat(samplingController.isInSample(neverInSample)).isFalse();
    }

}

File: analytics/testkitchen/src/test/java/org/wikimedia/testkitchen/SamplingControllerTest.kt
package org.wikimedia.testkitchen

import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Test
import org.wikimedia.testkitchen.config.StreamConfig
import org.wikimedia.testkitchen.config.sampling.SampleConfig
import org.wikimedia.testkitchen.context.DataFixtures

internal class SamplingControllerTest {
    private val samplingController = SamplingController(
        DataFixtures.getTestClientData(),
        SessionController()
    )

    @Test
    fun testGetSamplingValue() {
        val deviceVal = samplingController.getSamplingValue(SampleConfig.UNIT_DEVICE)
        assertTrue(deviceVal >= 0.0)
        assertTrue(deviceVal <= 1.0)
    }

    @Test
    fun testGetSamplingId() {
        assertNotNull(samplingController.getSamplingId(SampleConfig.UNIT_DEVICE))
        assertNotNull(samplingController.getSamplingId(SampleConfig.UNIT_SESSION))
    }

    @Test
    fun testNoSamplingConfig() {
        val noSamplingConfig = StreamConfig().also {
            it.streamName = "foo"
            it.schemaTitle = "bar"
        }
        assertTrue(samplingController.isInSample(noSamplingConfig))
    }

    @Test
    fun testAlwaysInSample() {
        val alwaysInSample = StreamConfig().also {
            it.streamName = "foo"
            it.schemaTitle = "bar"
            it.producerConfig = StreamConfig.ProducerConfig().also { pc ->
                pc.metricsPlatformClientConfig = StreamConfig.MetricsPlatformClientConfig()
            }
        }
        assertTrue(samplingController.isInSample(alwaysInSample))
    }

    @Test
    fun testNeverInSample() {
        val neverInSample = StreamConfig().also {
            it.streamName = "foo"
            it.schemaTitle = "bar"
            it.producerConfig = StreamConfig.ProducerConfig().also { pc ->
                pc.metricsPlatformClientConfig = StreamConfig.MetricsPlatformClientConfig()
            }
            it.sampleConfig = SampleConfig(0.0, SampleConfig.UNIT_SESSION)
        }
        assertFalse(samplingController.isInSample(neverInSample))
    }
}
